src/dpdkstat.c

```diff
@@ -0,0 +1,775 @@
+  if(g_configuration)
+    return 0;
+
+  /* Create and open a new object, or open an existing object. */
+  int fd = shm_open(DPDK_SHM_NAME, O_CREAT | O_TRUNC | O_RDWR, 0666);
+  if (fd < 0) {
+    WARNING("dpdkstat:Failed to open %s as SHM:%s\n", DPDK_SHM_NAME,
+            strerror(errno));
+    goto fail;
+  }
+  /* Set the size of the shared memory object. */
+  int ret = ftruncate(fd, size);
+  if (ret != 0) {
+    WARNING("dpdkstat:Failed to resize SHM:%s\n", strerror(errno));
+    goto fail_close;
+  }
+  /* Map the shared memory object into this process' virtual address space. */
+  g_configuration = mmap(0, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+  if (g_configuration == MAP_FAILED) {
+    WARNING("dpdkstat:Failed to mmap SHM:%s\n", strerror(errno));
+    goto fail_close;
+  }
+  /*
+   * Close the file descriptor, the shared memory object still exists
+   * and can only be removed by calling shm_unlink().
+   */
+  close(fd);
+
+  /* Initialize g_configuration. */
+  memset(g_configuration, 0, size);
+
+  /* Initialize the semaphores for SHM use */
+  int err = sem_init(&g_configuration->sema_helper_get_stats, 1, 0);
+  if(err) {
+    ERROR("dpdkstat semaphore init failed: %s\n", strerror(errno));
+    goto fail_close;
+  }
+  err = sem_init(&g_configuration->sema_stats_in_shm, 1, 0);
+  if(err) {
+    ERROR("dpdkstat semaphore init failed: %s\n", strerror(errno));
+    goto fail_close;
+  }
+
+  return 0;
+
+fail_close:
+  close(fd);
+fail:
+  /* Reset to zero, as it was set to MAP_FAILED aka: (void *)-1. Avoid
+   * an issue if collectd attempts to run this plugin failure.
+   */
+  g_configuration = 0;
+  return -1;
+}
+
+static int dpdk_re_init_shm()
+{
+  dpdk_config_t temp_config;
+  memcpy(&temp_config, g_configuration, sizeof(dpdk_config_t));
+  DEBUG("dpdkstat: %s: ports %d, xstats %d\n", __func__, temp_config.num_ports,
+        temp_config.num_xstats);
+
+  int shm_xstats_size = sizeof(dpdk_config_t) + (sizeof(struct rte_eth_xstats) *
+                        g_configuration->num_xstats);
+  DEBUG("=== SHM new size for %d xstats\n", g_configuration->num_xstats);
+
+  int err = dpdk_shm_cleanup();
+  if (err)
+    ERROR("dpdkstat: Error in shm_cleanup in %s\n", __func__);
+
+  err = dpdk_shm_init(shm_xstats_size);
+  if (err)
+    ERROR("dpdkstat: Error in shm_init in %s\n", __func__);
+
+  /* If the XML config() function has been run, dont re-initialize defaults */
+  if(!g_configured)
+    dpdk_config_init_default();
+
+  memcpy(g_configuration, &temp_config, sizeof(dpdk_config_t));
+  g_configuration->collectd_reinit_shm = 0;
+
+  return 0;
+}
+
+static int dpdk_init (void)
+{
+  int err = dpdk_shm_init(sizeof(dpdk_config_t));
+  if (err)
+    ERROR("dpdkstat: %s : error %d in shm_init()", __func__, err);
+
+  /* If the XML config() function has been run, dont re-initialize defaults */
+  if(!g_configured) {
+    dpdk_config_init_default();
+  }
+
+  plugin_register_complex_read (NULL, "dpdkstat", dpdk_read,
```

Comment:
I assume that you register the read callback here because you need to read `g_configuration->interval` from the config first. If you instead use the generic syntax:

```
<LoadPlugin "dpdkstat">
  Interval 42
</LoadPlugin>
```

and then pass `0` as interval, the daemon will automatically pick up the interval and you can move this call to `module_register()`.